### PR TITLE
Enhanced edit functionalities for group and key events 

### PR DESCRIPTION
**Pull request by sophiien**
**Eding Groups**

- To edit fields of a group, user must be in the module page. If not, there will be an error shown on the console to tell users to switch to the module page. 
- Format for editing groups: `edit INDEX o/group m/MODULE_CODE [g/NEW_GROUP_NAME] [mt/INDEX NEW_MEETING_TIME]`.  
- `INDEX` refers to the index of the group on the list of groups stated by the `m/MODULE_CODE`. The `INDEX` in `mt/` refers to the index of the meeting time to be edited. `INDEX`, `o/group` and `m/MODULE_CODE` are compulsory fields, and an error will be shown if either of these are missing. 
- [TODO] Editing person to add to a group function still not available, waiting for UI changes. 

**Editing Key Events**
- To edit fields of a key event, user must be in the module page. If not, there will be an error shown on the console to tell users to switch to the module page 
- Format for editing key events: `edit INDEX o/keyevent ke/INDEX [dt/DATETIME] [ty/TYPE]`. 
-  `INDEX` refers to the index of the module on the list of modules shown on the modules page. The `INDEX` in `ke/` refers to the index of the key event to be edited. `INDEX`, `o/keyevent` and `ke/INDEX` are compulsory fields, and an error will be shown if either of these are missing.

### CODE REVIEW

**Review by benmurphyy — CHANGES_REQUESTED** [review]
## Features missing, not following the standardised format set earlier.

- Refer to the shared document and follow the command format there. Implement missing commands.

**Comment by sophiien** [comment]
**Editing Groups**

- To edit fields of a group, user must be in the module or groups page. If not, there will be an error shown on the console to tell users to switch to the module / groups page.
- Format for editing groups: edit INDEX o/group m/MODULE_CODE [g/NEW_GROUP_NAME] [mt/INDEX NEW_MEETING_TIME].
- INDEX refers to the index of the group on the list of groups stated by the m/MODULE_CODE. The INDEX in mt/ refers to the index of the meeting time to be edited. INDEX, o/group and m/MODULE_CODE are compulsory fields, and an error will be shown if either of these are missing.
Screenshot 2022-03-29 at 12 23 19 AM


**Adding existing person to an existing group** 

- Format for editing person : edit INDEX o/person m/MODULE_CODE g/NEW_GROUP_NAME
- If any of the compulsory fields are missing, user will prompt them to use the correct format 
Screenshot 2022-03-29 at 12 16 01 AM
- If person at `INDEX` is a professor, will show an error that can only add students to a group 
Screenshot 2022-03-29 at 12 16 49 AM
- If person at `INDEX` is not a student stated by `m/MODULE_CODE`, will show error to first add the student before adding to group. 
Screenshot 2022-03-29 at 12 17 48 AM
- If person at `INDEX` already exists in the group, will show error that person is already in the group.
Screenshot 2022-03-29 at 12 18 34 AM


**Editing Key Events**

- To edit fields of a key event, user must be in the module page. If not, there will be an error shown on the console to tell users to switch to the module page
Screenshot 2022-03-29 at 12 34 16 AM

- Format for editing key events: edit INDEX o/keyevent ke/INDEX [dt/DATETIME] [type/TYPE].
- INDEX refers to the index of the module on the list of modules shown on the modules page. The INDEX in ke/ refers to the index of the key event to be edited. INDEX, o/keyevent and ke/INDEX are compulsory fields, and an error will be shown if either of these are missing.

**Review by benmurphyy — CHANGES_REQUESTED** [review]
## Missing features and wrong error messages
Screenshot 2022-03-29 at 11 18 42 AM
Screenshot 2022-03-29 at 11 20 44 AM

- remove o/ tag from command edits to maintain **app cohesion**

## Please follow documentation laid out to maintain app cohesion

**Review by benmurphyy — CHANGES_REQUESTED** [review]
## Follow documentation standardised format
- Confirm all the commands as stipulated in the document work before updating this PR.

File: src/main/java/unibook/logic/commands/EditCommand.java [hunk 1]
```diff
@@ -44,41 +54,81 @@
...
+            + "Existing values will be overwritten by the input values.\n"
+            + "Parameters: INDEX (of the module must be a positive integer) "
+            + PREFIX_OPTION + "OPTION "
+            + "[" + PREFIX_KEYEVENT + "KEYEVENT] "
+            + "[" + PREFIX_DATETIME + "DATETIME] \n "
+            + "Example: " + COMMAND_WORD + " 1 "
+            + PREFIX_OPTION + "keyevent "
+            + PREFIX_KEYEVENT + "1 "
+            + PREFIX_TYPE + "exam "
+            + PREFIX_MEETINGTIME + "12-12-2022 1645";
```
Comment (geraldneo567) [inline src/main/java/unibook/logic/commands/EditCommand.java:132]: One way to handle the lack of `option` is to have an enum type here, similar to what I have pasted in the screenshot.
image


Then when you create a command in `EditCommandParser`, you can pass in the respective type (removes the need for all the boolean checks, etc cos you know that the respective fields you need will not be null when you create a command with the specific enum. To handle different behaviour across different views, the `execute` method has booleans to show which group list is showing.

image


This is an example - see case `module` which is the same command that behaves differently on differnet pages.

**Review by geraldneo567 — CHANGES_REQUESTED** [review]

File: docs/UserGuide.md [hunk 2]
```diff
@@ -228,20 +228,63 @@ Format: `list mt/<YYYY-MM-DD>`
...
+
+Format: `edit INDEX o/keyevent ke/INDEX [type/TYPE] [dt/DATETIME]`
+* Edits  `type` and/or `datetime` of the keyevent at module at `INDEX`.
+* Example: `edit 1 o/keyevent ke/2 type/exam dt/2020-12-12 16:45` adds the second key event's type and date time in the first module in the list to exam and 12th December 2022 4:45pm respectively.
+
+**Commands while on `Groups` view**
+
+At least one optional field must be edited in order for module to be successfully edited.
+
+Format: `edit INDEX o/group m/MODULE [g/GROUPNAME] [mt/INDEX DATETIME]`
```
Comment (benmurphyy) [inline docs/UserGuide.md:284]: groupname is a compulsory field

**Review by benmurphyy — CHANGES_REQUESTED** [review]
- change the message shown when adding person to group or module to something like "added XX to group YYY" not this generic message
Screenshot 2022-03-29 at 8 27 29 PM

- unable to edit office of professor
Screenshot 2022-03-29 at 8 31 52 PM

**Comment by sophiien** [comment]
Fixed bug identified by @geraldneo567 in previous comment.

**Review by geraldneo567 — APPROVED** [review]

**Review by benmurphyy — APPROVED** [review]
LGTM

**Pull request merged**